src/Discord.Net.WebSocket/Entities/Invites/SocketInvite.cs

```diff
@@ -0,0 +1,133 @@
+using Discord.Rest;
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Model = Discord.API.Gateway.InviteCreateEvent;
+
+namespace Discord.WebSocket
+{
+    [DebuggerDisplay(@"{DebuggerDisplay,nq}")]
+    public class SocketInvite : SocketEntity<string>, IInviteMetadata
+    {
+        private long _createdAtTicks;
+
+        /// <inheritdoc />
+        public ulong ChannelId { get; private set; }
+        /// <summary>
+        ///     Gets the channel where this invite was created.
+        /// </summary>
+        public SocketGuildChannel Channel { get; private set; }
+        /// <inheritdoc />
+        public ulong? GuildId { get; private set; }
+        /// <summary>
+        ///     Gets the guild where this invite was created.
+        /// </summary>
+        public SocketGuild Guild { get; private set; }
+        /// <inheritdoc />
+        ChannelType IInvite.ChannelType
+        {
+            get
+            {
+                switch (Channel)
+                {
+                    case IVoiceChannel voiceChannel: return ChannelType.Voice;
+                    case ICategoryChannel categoryChannel: return ChannelType.Category;
+                    case IDMChannel dmChannel: return ChannelType.DM;
+                    case IGroupChannel groupChannel: return ChannelType.Group;
+                    case SocketNewsChannel socketNewsChannel: return ChannelType.News;
+                    case ITextChannel textChannel: return ChannelType.Text;
+                    default: throw new InvalidOperationException("Invalid channel type.");
+                }
+            }
+        }
+        /// <inheritdoc />
+        string IInvite.ChannelName => Channel.Name;
+        /// <inheritdoc />
+        string IInvite.GuildName => Guild.Name;
+        /// <inheritdoc />
+        int? IInvite.PresenceCount => throw new NotImplementedException();
+        /// <inheritdoc />
+        int? IInvite.MemberCount => throw new NotImplementedException();
+        /// <inheritdoc />
+        bool IInviteMetadata.IsRevoked => throw new NotImplementedException();
+        /// <inheritdoc />
+        public bool IsTemporary { get; private set; }
+        /// <inheritdoc />
+        int? IInviteMetadata.MaxAge { get => MaxAge; }
+        /// <inheritdoc />
+        int? IInviteMetadata.MaxUses { get => MaxUses; }
+        /// <inheritdoc />
+        int? IInviteMetadata.Uses { get => Uses; }
```

Comment:
1. Why are *these* properties explicitly implemented?
2. If they really should be, why the extra braces?

Reply:
They are required by IInviteMetadata, but shouldn't be optional for the socket entity according to the docs api ( https://discordapp.com/developers/docs/topics/gateway#invite-create ) so I added non-nullable ones and redirected these to them.

Reply:
But [the Invite Metadata object](https://discordapp.com/developers/docs/resources/invite#invite-metadata-object) also has them listed as not optional.
![](https://i.imgur.com/wWf3oBC.png)

Reply:
For some reason they were added as nullables, https://github.com/discord-net/Discord.Net/blob/dev/src/Discord.Net.Core/Entities/Invites/IInviteMetadata.cs#L39-L61
Not sure if it was changed later by discord or it's some workaround for the lib.

Also noticed our IInvite is missing two fields that were added to INVITE_CREATE and that it doesn't match discord, there's no `channel_name` or `channel_type` and `inviter` should be part of `Invite`, not `InviteMetadata` ( https://discordapp.com/developers/docs/resources/invite#invite-object ).

Might be remaining code trying to stay backward compatible, not sure.

Reply:
Seeing again, the `channel_name` and `channel_type` are just what discord sends and we use the cache to get the channel, since it also gets a partial channel unlike the gateway event.